File: 0807-max-increase-to-keep-city-skyline/0807-max-increase-to-keep-city-skyline.java
class Solution {
    public int maxIncreaseKeepingSkyline(int[][] grid) {
        int rows = grid.length;
        int cols = grid[0].length;

        int[] rowMax = new int[rows];
        int[] colMax = new int[cols];

        for (int r = 0; r < rows; r++) {
            int maxVal = 0;
            for (int c = 0; c < cols; c++) {
                if (grid[r][c] > maxVal) {
                    maxVal = grid[r][c];
                }
            }
            rowMax[r] = maxVal;
        }

        for (int c = 0; c < cols; c++) {
            int maxVal = 0;
            for (int r = 0; r < rows; r++) {
                if (grid[r][c] > maxVal) {
                    maxVal = grid[r][c];
                }
            }
            colMax[c] = maxVal;
        }

        int diff = 0;
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                int allowedHeight = Math.min(rowMax[r], colMax[c]);
                diff += (allowedHeight - grid[r][c]);
            }
        }
        return diff;
    }
}

File: 0807-max-increase-to-keep-city-skyline/0807-max-increase-to-keep-city-skyline.py
class Solution:
    def maxIncreaseKeepingSkyline(self, grid: List[List[int]]) -> int:
        
        rows = len(grid)
        cols = len(grid[0])

        new_grid = [[0] * cols for _ in range(rows)]
        max_rows = [0] * rows
        max_cols = [0] * cols
        for r in range(rows):
            for c in range(cols):
                max_rows[r] = max(max_rows[r], grid[r][c])
                max_cols[c] = max(max_cols[c], grid[r][c])
        
        print(max_rows)
        print(max_cols)

        for r in range(rows):
            for c in range(cols):
                new_grid[r][c] = min(max_rows[r], max_cols[c])

        print(new_grid)
        diff = 0
        for r in range(rows):
            for c in range(cols):
                diff += (new_grid[r][c] - grid[r][c])
        
        return diff

